helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -391,6 +406,34 @@ private static Set<AssignableReplica> findToBeAssignedReplicasByClusterChanges(
     return toBeAssignedReplicas;
   }
 
+  private static Set<AssignableReplica> findToBeAssignedReplicasIllegalPlacements(
+      Map<String, Set<AssignableReplica>> replicaMap, Set<String> activeInstances,
+      Map<String, ResourceAssignment> currentAssignment,
+      Map<String, Set<AssignableReplica>> allocatedReplicas) {
+    // For any replica tht does not exist in currentAssignment (new resources) or does not exist in
+    // active instances (down instances), return them.
+    Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+    for (String resourceName : replicaMap.keySet()) {
+      Map<String, Map<String, Set<String>>> currentPartitionStateMap =
+          getValidStateInstanceMap(currentAssignment.get(resourceName), activeInstances);
+      for (AssignableReplica replica : replicaMap.get(resourceName)) {
+        String partitionName = replica.getPartitionName();
+        String replicaState = replica.getReplicaState();
+        Set<String> currentAllocations =
+            currentPartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+        List<String> currentAllocationsList = new ArrayList<>(currentAllocations);
```

Comment:
i am sorry, but why are we creating this new List<>?

Reply:
The code tries to remove any element from the set, and the list is created to mimic the behavior on the partial rebalance side. However, I will find a better way to remove the element.